util/signing/signer/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+pub mod cryptography;
+
+/// A collection of bytes.
+#[derive(Debug, Clone)]
+pub struct Bytes(pub Vec<u8>);
+
+/// A signature.
+#[derive(Debug, Clone)]
+pub struct Signature(pub Bytes);
+
+/// A public key.
+#[derive(Debug, Clone)]
+pub struct PublicKey(pub Bytes);
+
+/// Version of a key.
+/// Default mean the current key.
+#[derive(Debug, Clone, Default)]
+pub struct KeyVersion(pub String);
+
+/// Id that identify a Key.
+#[derive(Debug, Clone)]
+pub struct KeyId(pub String);
+
+/// Errors thrown by Signer
+#[derive(Debug, thiserror::Error)]
+pub enum SignerError {
+	#[error("Error during signing : {0}")]
+	Sign(String),
+	#[error("Error during public key retrieval : {0}")]
+	PublicKey(String),
+	#[error("Error can't decode provided hex data : {0}")]
+	Hex(String),
+	#[error("Signature not found.")]
+	SignatureNotFound,
+	#[error("public key not found.")]
+	PublicKeyNotFound,
+}
+
+#[async_trait::async_trait]
+pub trait SignerOperations<C: cryptography::Curve> {
+	/// Signs some bytes.
+	async fn sign(&self, message: Bytes) -> Result<Signature, SignerError>;
+
+	/// Gets the public key.
+	async fn public_key(&self) -> Result<PublicKey, SignerError>;
+}
+
+pub struct Signer<O, C>
+where
+	O: SignerOperations<C>,
+	C: cryptography::Curve,
```

Comment:
I don't see why you define a generic struct for Signer that do the same as SignerOperation. It's a wrapper. Perhaps each real Signer should implement the trait and we use the trait as function parameters.

Reply:
See my comment on @mzabaluev question about the same.